api/services/passport.js

```diff
@@ -1,85 +1,92 @@
 const GitHubStrategy = require('passport-github').Strategy;
 const Passport = require('passport');
 const config = require('../../config');
-const { User, Event } = require('../models');
 const { logger } = require('../../winston');
+const { User, Event } = require('../models');
 const GitHub = require('./GitHub');
 const RepositoryVerifier = require('./RepositoryVerifier');
 const EventCreator = require('./EventCreator');
+const { createUAAStrategy } = require('./uaaStrategy');
 
 const passport = new Passport.Passport();
 
-const githubVerifyCallback = (accessToken, refreshToken, profile, callback) => {
-  let user;
-
-  return GitHub.validateUser(accessToken)
-    .then(() => User.findOrCreate({
-      where: { username: profile.username.toLowerCase() },
-      defaults: {
-        // eslint-disable-next-line no-underscore-dangle
-        email: profile._json.email,
-        username: profile.username,
-      },
-    }))
-    .then((models) => {
-      [user] = models;
-      if (!user) {
-        throw new Error(`Unable to find or create user ${profile.username}`);
-      }
-      const eventBody = { action: 'login' };
-      EventCreator.audit(Event.labels.AUTHENTICATION, user, eventBody);
-
-      return user.update({
-        githubAccessToken: accessToken,
-        githubUserId: profile.id,
-        signedInAt: new Date(),
-      });
-    })
-    .then(() => {
-      RepositoryVerifier.verifyUserRepos(user); // verify user's site's repos
-      callback(null, user);
-    })
-    .catch((err) => {
-      logger.warn('Authentication error: ', err);
-      callback(err);
+const {
+  github: { options: githubOptions },
+  uaa: { options: uaaOptions },
+} = config.passport;
+
+async function verifyGithub(accessToken, _refreshToken, profile, callback) {
+  try {
+    const isValidUser = await GitHub.validateUser(accessToken, false);
+    if (!isValidUser) {
+      return callback(null, false);
+    }
+
+    const username = profile.username.toLowerCase();
+    // eslint-disable-next-line no-underscore-dangle
+    const { email } = profile._json;
+
+    const [user] = await User.upsert({
+      username,
+      email,
+      githubAccessToken: accessToken,
+      githubUserId: profile.id,
+      signedInAt: new Date(),
     });
-};
 
-passport.use(new GitHubStrategy(config.passport.github.options, githubVerifyCallback));
+    EventCreator.audit(Event.labels.AUTHENTICATION, user, { action: 'login' });
 
-passport.logout = (req, res) => {
-  const { user } = req;
-  req.logout();
-  if (user) {
-    const eventBody = { action: 'logout' };
-    EventCreator.audit(Event.labels.AUTHENTICATION, user, eventBody);
+    // should we wait for this?
+    RepositoryVerifier.verifyUserRepos(user);
```

Comment:
should we wait for this to finish?

Reply:
i don't have a strong opinion on this but I lean towards not waiting.   i mainly see the benefit in waiting for a user who has not logged in for quite a while